### PR TITLE
fix(api): support multiple belongsTo

### PR DESCRIPTION
**Pull request by fjnoyp**
*Issue #, if available:* https://github.com/aws-amplify/amplify-flutter/issues/1814

*Description of changes:* Support multiple belongsTo

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by dnys1** [comment]
Can we add some tests?

**Comment by HuiSF** [comment]
Take schema `many posts to many tags` - Please add tests verifying selection set for:

1. query/mutate a post, result should include tags associated with the post
2. query/mutate a tag, result should include posts associated with the tag
3. query/mutate `PostTags` the intermediate model of the many to many relationship, result should include all associated posts and tags.

File: packages/api/amplify_api/lib/src/graphql/utils.dart [hunk 1]
```diff
@@ -51,8 +51,8 @@ _RelatedFields _getRelatedFields(ModelSchema modelSchema) {
   return _fieldsMemo[modelSchema]!;
 }
 
-ModelField? getBelongsToFieldFromModelSchema(ModelSchema modelSchema) {
-  return _getRelatedFields(modelSchema).singleFields.firstWhereOrNull((entry) =>
+Iterable<ModelField> getBelongsToFieldFromModelSchema(ModelSchema modelSchema) {
```
Comment (ragingsquirrel3) [inline packages/api/amplify_api/lib/src/graphql/utils.dart:54]: nit: could you please rename this `getBelongsToFieldsFromModelSchema` so it's clear it returns multiple.

**Review by ragingsquirrel3 — COMMENTED** [review]
Change looks good, like others said, just waiting to see tests.

**Comment by fjnoyp** [comment]
Tests added. 

@HuiSF for the cases you mentioned for manyToMany wouldn't the base models themselves not contain the relationships?  You'd have to query on the intermediate object instead right?

**Comment by HuiSF** [comment]
> Tests added. 
> 
> 
> 
> @HuiSF for the cases you mentioned for manyToMany wouldn't the base models themselves not contain the relationships?  You'd have to query on the intermediate object instead right?

No the way AppSync works is when you query either the end model, it returns related models via the connecting intermediate model under the hood. You can try it out in AppSync console to see how it works (with the selection set). I think directly operating the intermediate model is only necessary with DataStore use cases (e.g. creating many-to-many relationship) I may be wrong, but you should be able to verify playing with AppSync mutation easily in the AppSync console.

**Comment by MarlonJD** [comment]
> > Tests added. 
> > 
> > 
> > 
> > @HuiSF for the cases you mentioned for manyToMany wouldn't the base models themselves not contain the relationships?  You'd have to query on the intermediate object instead right?
> 
> No the way AppSync works is when you query either the end model, it returns related models via the connecting intermediate model under the hood. You can try it out in AppSync console to see how it works (with the selection set). I think directly operating the intermediate model is only necessary with DataStore use cases (e.g. creating many-to-many relationship) I may be wrong, but you should be able to verify playing with AppSync mutation easily in the AppSync console.

Main issue was the malformed content and removed 2nd belongsTo issue. If we can see the right graphql query/mutation and variable on document. İt should work under AppSync console. But there was a sync issue according to this belongsTo issue, i hope this pr also solve that issue. I'm so happy that you are working with this issue, i was waiting for this for 3 months.

**Comment by MarlonJD** [comment]
> Tests added. 
> 
> @HuiSF for the cases you mentioned for manyToMany wouldn't the base models themselves not contain the relationships?  You'd have to query on the intermediate object instead right?

Yes, if it's manyToMany, it's creating new model as you said -intermediate object- You can directly query to this object. İt should have 2 belongsTo. And there was a certain creation and getting issue on datastore and API. Also it has been sync issue if manyToMany on datastore, i think sync issue will resolved after this fix. I'm so happy that change becoming through. Thanks man

**Review by dnys1 — APPROVED** [review]

**Review by dnys1 — APPROVED** [review]

File: packages/api/amplify_api/test/graphql_helpers_test.dart [hunk 2]
```diff
@@ -40,8 +41,118 @@ void main() {
...
+          id: 'firstMtmRelationId',
+          manyToManyPrimary: primary,
+          manyToManySecondary: secondary);
+      GraphQLRequest<FirstMtmRelation> req =
+          ModelMutations.update<FirstMtmRelation>(relation);
+
+      expect(req.document, expectedDoc);
+      expect(_deepEquals(req.variables, expectedVars), isTrue);
+      expect(req.modelType, FirstMtmRelation.classType);
+      expect(req.decodePath, 'updateFirstMtmRelation');
```
Comment (ragingsquirrel3) [inline packages/api/amplify_api/test/graphql_helpers_test.dart:149]: request: Could you please add a test that the decoder handles this situation as expected? See below examples with mocked server response. Will need a mocked server response (dart map) a like AppSync would provide in this scenario passed directly to decoder and that resulting response has model with relationships populated as expected.

**Review by ragingsquirrel3 — APPROVED** [review]

**Review by ragingsquirrel3 — APPROVED** [review]

**Review by dnys1 — APPROVED** [review]

**Comment by ignitum-ryan** [comment]
@fjnoyp - Any plans on GA release of this feature so we can pull it in and begin testing? (Attempted to pull in the main branch rather than a GA version, but was informed there is no pubspec.yaml in the repo for pubspec to pull it in.)

**Comment by MarlonJD** [comment]
> @fjnoyp - Any plans on GA release of this feature so we can pull it in and begin testing? (Attempted to pull in the main branch rather than a GA version, but was informed there is no pubspec.yaml in the repo for pubspec to pull it in.)

I'm using this patch like this: I changed 2 files for local copy of my pub package of amplify-flutter. 
([packages/api/amplify_api/lib/src/graphql/utils.dart](https://github.com/aws-amplify/amplify-flutter/pull/2087#diff-c24315a30cd09bc2d990b8804008fa97b22c2d1f88288c7b08bc8489b7f4fb2b))
and
[packages/api/amplify_api/lib/src/graphql/graphql_request_factory.dart](https://github.com/aws-amplify/amplify-flutter/pull/2087#diff-9cfd63722dfdc451e4d04281d279bf95f17cfd7980e27f279c96794c63eedb6b)

Copy all files and replace your local copy and use it, but I think It will be on next release, but I dont know when.

Also you can use [amplify-flutter-1 of fjnoyp](https://github.com/fjnoyp/amplify-flutter-1 and use this ref fb25ae1154dea5f187a10169f318b798ffb15bbc. for example

```
amplify_flutter:
    git: 
      url: https://github.com/fjnoyp/amplify-flutter-1
      ref: fb25ae1154dea5f187a10169f318b798ffb15bbc
```

**Comment by fjnoyp** [comment]
Hi @ignitum-ryan  @MarlonJD  thank you for your comments.  I'll check internally when our next release and let you know, so you don't need to make these workarounds.

**Comment by ignitum-ryan** [comment]
> 

Thanks so much @MarlonJD for the details. I may need to do this in the near future, as well as others, so the details are appreciated!

**Comment by ignitum-ryan** [comment]
> Hi @ignitum-ryan @MarlonJD thank you for your comments. I'll check internally when our next release and let you know, so you don't need to make these workarounds.

Thanks for checking in on this! Excited to see this get released as it's going to open a lot of opportunity for Amplify users.

**Pull request merged**